### PR TITLE
gh-106368: Increase Argument Clinic CLI test coverage

### PR DESCRIPTION
**Pull request by erlend-aasland**
Instead of hacking into the Clinic class, use the Argument Clinic tool
to run the ~ClinicWholeFileTest~ ClinicExternalTest test suite.

TODO:

- [x] Remove tmp stuff from `test_external`, use `-o` to redirect the output, and simply compare the redirected output with `clinic.test.c`
- [x] A new test that verifies that the file is not touched if no changes are made (currently a part of `test_external`)
- [x] A test that verifies that an incorrect output checksum (and the generated output) is regenerated if `-f` is given
- [x] Test verbose mode

* Issue: gh-106368

### CODE REVIEW

**Comment by erlend-aasland** [comment]
FWIW; this messes up test coverage locally; that's solvable (IIRC), though.

**Comment by erlend-aasland** [comment]
The coveragerc file that's included in the sources disables coverage for `if __name__ == "__main__"`, which is unfortunate; we should change that.

File: Lib/test/test_clinic.py [hunk 1]
```diff
@@ -1346,31 +1349,182 @@ def test_scaffolding(self):
...
+
+            All converters also accept (c_default=None, py_default=None, annotation=None).
+            All return converters also accept (py_default=None).
+        """)
+        out = self.expect_success("--converters")
+        # We cannot simply compare the output, because the repr of the *accept*
+        # param may change (it's a set, thus unordered). So, let's compare the
+        # start and end of the expected output, and then assert we've got a
+        # line for each converter.
+        self.assertTrue(out.startswith(prelude))
```
Comment (sobolevn) [inline Lib/test/test_clinic.py]:
```suggestion
        self.assertTrue(out.startswith(prelude), out)
```

File: Lib/test/test_clinic.py [hunk 2]
```diff
@@ -1346,31 +1349,182 @@ def test_scaffolding(self):
...
+        for converter in expected_converters:
+            with self.subTest(converter=converter):
+                self.assertIn(converter, out)
+        self.assertTrue(out.endswith(finale))
```
Comment (sobolevn) [inline Lib/test/test_clinic.py]:
```suggestion
        self.assertTrue(out.endswith(finale), out)
```

**Review by sobolevn — COMMENTED** [review]
Better error messages :)

**Comment by erlend-aasland** [comment]
Thanks for the review, Nikita!

**Review by sobolevn — APPROVED** [review]
👍

File: Lib/test/test_clinic.py [hunk 3]
```diff
@@ -1346,31 +1349,182 @@ def test_scaffolding(self):
...
+        """)
+        out = self.expect_success("--converters")
+        # We cannot simply compare the output, because the repr of the *accept*
+        # param may change (it's a set, thus unordered). So, let's compare the
+        # start and end of the expected output, and then assert we've got a
+        # line for each converter.
+        self.assertTrue(out.startswith(prelude), out)
+        for converter in expected_converters:
+            with self.subTest(converter=converter):
+                self.assertIn(converter, out)
```
Comment (erlend-aasland) [inline Lib/test/test_clinic.py]: We could make this part slightly more accurate. For example, break the output into an array of lines, lstrip each line and assert with str.startswith.

Comment (erlend-aasland) [reply Lib/test/test_clinic.py]: What do you think about a4adb705d7d8e23d413a8f12f7066df558d10ca0, @sobolevn?

**Review by sobolevn — APPROVED** [review]
Looks great! 👍

File: Lib/test/test_clinic.py (hunk 3 above)
Comment (sobolevn) [reply Lib/test/test_clinic.py]: Yes, I think is better now. Order is important.

**Pull request merged**